File: daemon/src/ingest/post_tool_use.rs
//! `POST /hooks/post-tool-use` handler.
//!
//! Appends a `tool_result` timeline entry (never changes status —
//! `session_state::transition` maps `PostToolUse` to a no-op) and clears
//! `current_tool` (the in-flight tool call has finished). If the payload
//! indicates the tool call itself failed, the event is recorded as an
//! `error` (D-10/MON-05: visible in the timeline only, never a status
//! change or notification — `PostToolUse` was already a status no-op, so
//! this is consistent either way).

use axum::{extract::State, http::StatusCode, Json};
use serde_json::Value;
use std::sync::Arc;

use crate::{ingest, session_state::HookEvent, AppState, IngestEventRequest};

pub async fn post_tool_use(
    State(state): State<Arc<AppState>>,
    Json(payload): Json<Value>,
) -> Result<StatusCode, StatusCode> {
    let session_id = payload
        .get("session_id")
        .and_then(Value::as_str)
        .ok_or(StatusCode::BAD_REQUEST)?
        .to_string();
    let cwd = payload
        .get("cwd")
        .and_then(Value::as_str)
        .map(str::to_string);
    let tool_name = payload
        .get("tool_name")
        .and_then(Value::as_str)
        .map(str::to_string);
    let condensed_output = payload
        .get("tool_response")
        .map(|v| ingest::condensed_json_summary(v, 200))
        .unwrap_or_default();
    let summary = match &tool_name {
        Some(name) => format!("{name}: {condensed_output}"),
        None => condensed_output,
    };

    // Best-effort detection of a failed tool call — exact field name/shape
    // unconfirmed (RESEARCH.md payload-uncertainty note); check the common
    // candidates defensively rather than assuming one.
    let is_error = payload
        .get("tool_response")
        .and_then(|r| r.get("is_error"))
        .and_then(Value::as_bool)
        .or_else(|| payload.get("is_error").and_then(Value::as_bool))
        .unwrap_or(false);

    let request = IngestEventRequest {
        session_id,
        event: HookEvent::PostToolUse,
        cwd,
        notification_type: None,
        tool_name,
        timeline_summary: summary,
        payload_json: Some(payload.to_string()),
        is_error,
        first_prompt_text: None,
        mark_ended: false,
    };

    ingest::dispatch_ingest_event(&state, request).await
}

File: daemon-ts/src/ingest/postToolUse.ts
/**
 * `POST /hooks/post-tool-use` handler.
 *
 * Appends a `tool_result` timeline entry (never changes status —
 * `sessionState.transition` maps `PostToolUse` to a no-op) and clears
 * `current_tool` (the in-flight tool call has finished). If the payload
 * indicates the tool call itself failed, the event is recorded as an
 * `error` (MON-05: visible in the timeline only, never a status change or
 * notification — `PostToolUse` was already a status no-op, so this is
 * consistent either way). Mirrors `daemon/src/ingest/post_tool_use.rs`.
 */

import type { FastifyReply, FastifyRequest } from "fastify";
import type { Database as DatabaseType } from "better-sqlite3";

import { condensedJsonSummary } from "../store.js";
import { dispatchIngestEvent } from "./dispatch.js";
import type { IngestEventRequest } from "./dispatch.js";

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export function makePostToolUseHandler(db: DatabaseType) {
  return async (req: FastifyRequest, reply: FastifyReply): Promise<void> => {
    const body = (req.body ?? {}) as Record<string, unknown>;
    const sessionId = typeof body.session_id === "string" ? body.session_id : undefined;
    if (!sessionId) {
      reply.code(400).send();
      return;
    }
    const cwd = typeof body.cwd === "string" ? body.cwd : null;
    const toolName = typeof body.tool_name === "string" ? body.tool_name : null;
    const toolResponse = body.tool_response;
    const condensedOutput = "tool_response" in body ? condensedJsonSummary(toolResponse, 200) : "";
    const summary = toolName ? `${toolName}: ${condensedOutput}` : condensedOutput;

    // Best-effort detection of a failed tool call — checks
    // tool_response.is_error first, then the top-level is_error field.
    const nestedIsError = isRecord(toolResponse) ? toolResponse.is_error : undefined;
    const isError =
      (typeof nestedIsError === "boolean" ? nestedIsError : undefined) ??
      (typeof body.is_error === "boolean" ? body.is_error : undefined) ??
      false;

    const request: IngestEventRequest = {
      sessionId,
      event: "PostToolUse",
      cwd,
      notificationType: null,
      toolName,
      timelineSummary: summary,
      payloadJson: JSON.stringify(body),
      isError,
      firstPromptText: null,
      markEnded: false,
    };

    dispatchIngestEvent(db, request);
    reply.code(200).send();
  };
}
